File: libs/types/types/services/twitch/chat-message.go
package twitch

const TOPIC_CHAT_MESSAGE = "twitch.chat_message"

type TwitchChatMessage struct {
	BroadcasterUserId           string
	BroadcasterUserName         string
	BroadcasterUserLogin        string
	ChatterUserId               string
	ChatterUserName             string
	ChatterUserLogin            string
	MessageId                   string
	Message                     *ChatMessageMessage
	Color                       string
	Badges                      []ChatMessageBadge
	MessageType                 string
	Cheer                       *ChatMessageCheer
	Reply                       *ChatMessageReply
	ChannelPointsCustomRewardId string
}

type FragmentType int32

const (
	FragmentType_TEXT      FragmentType = 0
	FragmentType_CHEERMOTE FragmentType = 1
	FragmentType_EMOTE     FragmentType = 2
	FragmentType_MENTION   FragmentType = 3
)

type ChatMessageMessageFragmentEmote struct {
	Id         string
	EmoteSetId string
	OwnerId    string
	Format     []string
}

type ChatMessageMessageFragmentMention struct {
	UserId    string
	UserName  string
	UserLogin string
}

type ChatMessageMessageFragmentCheermote struct {
	Prefix string
	Bits   int64
	Tier   int64
}

type ChatMessageMessageFragment struct {
	Type      FragmentType
	Text      string
	Cheermote *ChatMessageMessageFragmentCheermote
	Emote     *ChatMessageMessageFragmentEmote
	Mention   *ChatMessageMessageFragmentMention
}

type ChatMessageMessage struct {
	Text      string
	Fragments []ChatMessageMessageFragment
}

type ChatMessageBadge struct {
	Id    string
	SetId string
	Info  string
}

type ChatMessageCheer struct {
	Bits int64
}

type ChatMessageReply struct {
	ParentMessageId   string
	ParentMessageBody string
	ParentUserId      string
	ParentUserName    string
	ParentUserLogin   string
	ThreadMessageId   string
	ThreadUserId      string
	ThreadUserName    string
	ThreadUserLogin   string
}

File: libs/bus-core/src/generic/chat-message.ts
export interface ChatMessageBadge {
	readonly id?: string
	readonly set_id: string
	readonly info?: string
	readonly text: string
}

export interface ChatMessageFragment {
	readonly text: string
	readonly position?: {
		readonly start?: number
		readonly end?: number
	}
	readonly type: number
}

export interface ChatMessage {
	readonly message?: {
		readonly text: string
		readonly fragments?: readonly ChatMessageFragment[]
	}
	readonly id?: string
	readonly broadcaster_user_id?: string
	readonly broadcaster_user_name?: string
	readonly broadcaster_user_login?: string
	readonly chatter_user_id?: string
	readonly chatter_user_name?: string
	readonly chatter_user_login?: string
	readonly message_type?: string
	readonly platform: string
	readonly channel_id: string
	readonly channel_binding_id: string
	readonly user_id: string
	readonly platform_channel_id: string
	readonly sender_id: string
	readonly sender_login: string
	readonly sender_display_name: string
	readonly message_id: string
	readonly text: string
	readonly badges?: readonly ChatMessageBadge[]
	readonly is_broadcaster: boolean
	readonly is_moderator: boolean
	readonly is_vip: boolean
	readonly is_subscriber: boolean
	readonly color: string
}
